cylc/flow/etc/job.sh

```diff
@@ -121,11 +119,16 @@ cylc__job__main() {
     mkdir -p "$(dirname "${CYLC_TASK_WORK_DIR}")" || true
     mkdir -p "${CYLC_TASK_WORK_DIR}"
     cd "${CYLC_TASK_WORK_DIR}"
-    # User Environment, Pre-Script, Script and Post-Script
-    typeset func_name=
-    for func_name in 'user_env' 'pre_script' 'script' 'post_script'; do
-        cylc__job__run_inst_func "${func_name}"
-    done
+    # Env-Script, User Environment, Pre-Script, Script and Post-Script
+    # Execute in the subshell to protect job script environment from
+    # interferance and put in the background to allow signal traps to trigger.
+    (
+        typeset func_name=
+        for func_name in 'env_script' 'user_env' 'pre_script' 'script' 'post_script'; do
+            cylc__job__run_inst_func "${func_name}"
+        done
+    ) &
+    wait $!
```

Comment:
@TomekTrzeciak I am probably the one with less Shell knowledge in the team, but had some spare time and decided to have a look - as I know others are busy at the moment too.

Checked out your branch, then chose one of the failing tests for testing: `tests/cylc-poll/11-event-time.t`. It fails to run the suite reference check.

To identify what could have caused it, and as your PR is changing only a few lines of a single file, I went "toggling" your changes (i.e. reverting parts of it).

This line seems to cause the issue for this test on my environment. From what I understand, we start a subshell with the `(` and `)`. Send it to the background with `&`, then wait until this process is done with `wait $!`.

It looks good to me. But the test passes only when waiting for all the processes with `wait` (without the PID) instead of `wait $`. Not really sure why? :confused:

Reply:
Many thanks for looking into this @kinow. These tests will not work without modification as they rely on modifying/triggering traps in ways that are not accessible from a subshell/background process. However, I'm almost done with fixing all the failing tests. The key idea is to use `init-script` for setting up the trap related stuff, since this script is executed in the same shell as the job script.

Reply:
Will re-run the tests once you've finished working on this branch then @TomekTrzeciak. At least I am learning a bit more of Shell reading your changes.